### ISSUE TITLE
3D variable can't be read correctly

### ISSUE THREAD

**Issue by MartinKielhornPhilips** [batchatco/go-native-netcdf#6]
TL;DR I realized that this library does not allow to read NetCDF variables with more than two dimensions. I really would like to have support for three dimensions and developed a patch. I even created patch to support three to seven dimensions but including is a lot of code and I could not properly test it.

I try to write and read back a 3D array.

The written array dump looks okay:
```
~/ $ ncdump newdata.nc 
netcdf newdata {
dimensions:
        x = 2 ;
        y = 3 ;
        z = 4 ;
variables:
        ubyte x(x) ;
        ubyte y(y) ;
        ubyte z(z) ;
        ushort val(x, y, z) ;

// global attributes:
data:

 x = 0, 1 ;

 y = 0, 1, 2 ;

 z = 0, 1, 2, 3 ;

 val =
  0, 1, 2, 3,
  10, 11, 12, 13,
  20, 21, 22, 23,
  100, 101, 102, 103,
  110, 111, 112, 113,
  120, 121, 122, 123 ;
}
```
Note that each data value in the variable 'val' occurs only once.

The problem is with the data that is read back. My test program prints the following like:

```
2022-08-03 12:35:17.758  val[0]=[[0 1 2 3] [10 11 12 13] [20 21 22 23]] val[((len(val))-(1))]=[[3 10 11 12] [13 20 21 22] [23 100 101 102]] len(val)=2
```
Note that values from val[0] in val[1] repeat. This is wrong I expect it to look like that:

```
2022-08-03 12:35:17.758  val[0]=[[0 1 2 3] [10 11 12 13] [20 21 22 23]] val[((len(val))-(1))]=[[100 101 102 103] [110 111 112 113] [120 121 122 123]] len(val)=2
```

My test code is this. It generates the file newdata.nc using go-native-netcdf and tries to read it back:

```go
package main

import (
	"fmt"
	"github.com/batchatco/go-native-netcdf/netcdf/api"
	"github.com/batchatco/go-native-netcdf/netcdf/cdf"
	"github.com/samber/lo"
	"runtime"
	"time"
)

func timeNow() string {
	return time.Now().Format("2006-01-02 15:04:05.000")
}
func writeXarray(fn string) {
	cw, err00 := cdf.OpenWriter(fn)
	if !((err00) == (nil)) {
		fmt.Printf("%v cdf.OpenWriter(fn) err00=%v\n", timeNow(), err00)
		panic(err00)
	}
	fmt.Printf("%v define coordinate x \n", timeNow())
	nx := 2
	x := lo.Map[int, uint8](lo.Range(2), func(x int, _ int) uint8 {
		return uint8(x)
	})
	err01 := cw.AddVar("x", api.Variable{x, []string{"x"}, nil})
	if !((err01) == (nil)) {
		fmt.Printf("%v cw.AddVar('x', api.Variable {x, []string {'x'}, nil}) err01=%v\n", timeNow(), err01)
		panic(err01)
	}
	fmt.Printf("%v define coordinate y \n", timeNow())
	ny := 3
	y := lo.Map[int, uint8](lo.Range(3), func(x int, _ int) uint8 {
		return uint8(x)
	})
	err02 := cw.AddVar("y", api.Variable{y, []string{"y"}, nil})
	if !((err02) == (nil)) {
		fmt.Printf("%v cw.AddVar('y', api.Variable {y, []string {'y'}, nil}) err02=%v\n", timeNow(), err02)
		panic(err02)
	}
	fmt.Printf("%v define coordinate z \n", timeNow())
	nz := 4
	z := lo.Map[int, uint8](lo.Range(4), func(x int, _ int) uint8 {
		return uint8(x)
	})
	err03 := cw.AddVar("z", api.Variable{z, []string{"z"}, nil})
	if !((err03) == (nil)) {
		fmt.Printf("%v cw.AddVar('z', api.Variable {z, []string {'z'}, nil}) err03=%v\n", timeNow(), err03)
		panic(err03)
	}
	fmt.Printf("%v fill 3d dataset val \n", timeNow())
	val := make([][][]uint16, nx)
	for i := 0; i < nx; (i)++ {
		val[i] = make([][]uint16, ny)
		for j := 0; j < ny; (j)++ {
			val[i][j] = make([]uint16, nz)
			for k := 0; k < nz; (k)++ {
				val[i][j][k] = uint16((((100) * (i)) + ((10) * (j)) + ((1) * (k))))
			}
		}
	}
	err04 := cw.AddVar("val", api.Variable{val, []string{"x", "y", "z"}, nil})
	if !((err04) == (nil)) {
		fmt.Printf("%v cw.AddVar('val', api.Variable {val, []string {'x', 'y', 'z'}, nil}) err04=%v\n", timeNow(), err04)
		panic(err04)
	}
	defer (func() {
		fmt.Printf("%v close fn=%v\n", timeNow(), fn)
		err05 := cw.Close()
		if !((err05) == (nil)) {
			fmt.Printf("%v cw.Close() err05=%v\n", timeNow(), err05)
			panic(err05)
		}
	})()
}
func readXarray(fn string) {
	nc, err06 := cdf.Open(fn)
	if !((err06) == (nil)) {
		fmt.Printf("%v cdf.Open(fn) err06=%v\n", timeNow(), err06)
		panic(err06)
	}
	defer (func() {
		fmt.Printf("%v close file fn=%v\n", timeNow(), fn)
		nc.Close()
	})()
	fmt.Printf("%v obtain and show variable 'x' from netcdf file \n", timeNow())
	vrX, err07 := nc.GetVariable("x")
	if !((err07) == (nil)) {
		fmt.Printf("%v nc.GetVariable('x') err07=%v\n", timeNow(), err07)
		panic(err07)
	}
	if (vrX) == (nil) {
		fmt.Printf("%v variable x not found \n", timeNow())
		panic("")
	}
	x, hasX := vrX.Values.([]uint8)
	if !(hasX) {
		fmt.Printf("%v variable x conversion to []uint8 failed \n", timeNow())
		panic("")
	}
	fmt.Printf("%v  x[0]=%v x[((len(x))-(1))]=%v len(x)=%v\n", timeNow(), x[0], x[((len(x))-(1))], len(x))
	fmt.Printf("%v obtain and show variable 'y' from netcdf file \n", timeNow())
	vrY, err08 := nc.GetVariable("y")
	if !((err08) == (nil)) {
		fmt.Printf("%v nc.GetVariable('y') err08=%v\n", timeNow(), err08)
		panic(err08)
	}
	if (vrY) == (nil) {
		fmt.Printf("%v variable y not found \n", timeNow())
		panic("")
	}
	y, hasY := vrY.Values.([]uint8)
	if !(hasY) {
		fmt.Printf("%v variable y conversion to []uint8 failed \n", timeNow())
		panic("")
	}
	fmt.Printf("%v  y[0]=%v y[((len(y))-(1))]=%v len(y)=%v\n", timeNow(), y[0], y[((len(y))-(1))], len(y))
	fmt.Printf("%v obtain and show variable 'z' from netcdf file \n", timeNow())
	vrZ, err09 := nc.GetVariable("z")
	if !((err09) == (nil)) {
		fmt.Printf("%v nc.GetVariable('z') err09=%v\n", timeNow(), err09)
		panic(err09)
	}
	if (vrZ) == (nil) {
		fmt.Printf("%v variable z not found \n", timeNow())
		panic("")
	}
	z, hasZ := vrZ.Values.([]uint8)
	if !(hasZ) {
		fmt.Printf("%v variable z conversion to []uint8 failed \n", timeNow())
		panic("")
	}
	fmt.Printf("%v  z[0]=%v z[((len(z))-(1))]=%v len(z)=%v\n", timeNow(), z[0], z[((len(z))-(1))], len(z))
	fmt.Printf("%v obtain and show variable 'val' from netcdf file \n", timeNow())
	vrVal, err10 := nc.GetVariable("val")
	if !((err10) == (nil)) {
		fmt.Printf("%v nc.GetVariable('val') err10=%v\n", timeNow(), err10)
		panic(err10)
	}
	if (vrVal) == (nil) {
		fmt.Printf("%v variable val not found \n", timeNow())
		panic("")
	}
	val, hasVal := vrVal.Values.([][][]uint16)
	if !(hasVal) {
		fmt.Printf("%v variable val conversion to [][][]uint16 failed \n", timeNow())
		panic("")
	}
	fmt.Printf("%v  val[0]=%v val[((len(val))-(1))]=%v len(val)=%v\n", timeNow(), val[0], val[((len(val))-(1))], len(val))
}
func main() {
	fmt.Printf("%v netcdf runtime.Version()=%v\n", timeNow(), runtime.Version())
	fn := "newdata.nc"
	writeXarray(fn)
	readXarray(fn)
}

```

**Comment by MartinKielhornPhilips** [comment]
I did try reading the generated netCDF file with Python and github.com/fhs/go-netcdf. That works, so I'm pretty sure the generated file is good.

I really would like to be able to a read the file with a statically linked portable application using your library.

**Comment by MartinKielhornPhilips** [comment]
I started stepping through the code. In cdf.go:742  
the variable `data` contains the entire dataset as 1D array.

```go
		err := binary.Read(bf, binary.BigEndian, data)
		
		[]uint16 len: 24, cap: 24, [0,1,2,3,10,11,12,13,20,21,22,23,100,101,102,103,110,111,112,113,120,121,122,123]
```

In line 750 the variable converted contains repeated values:
```go
		converted := cdf.convert(data, dimLengthsCopy, varFound.vType)

interface {}([][][]uint16) [[[0,1,2,3],[10,11,12,13],[20,21,22,23]],[[3,10,11,12],[13,20,21,22],[23,100,101,102]]]
```
So I think the culprit is the function cdf.convert.

**Comment by MartinKielhornPhilips** [comment]
I fixed the issue by adding the following code after cdf.go:1082:

```go
if !cdf.slowConvert && len(dimLengths) == 3 {
		switch vType {
	

		case typeUShort:
			fmt.Println("!slowConvert len(dimLengths)==3.")
			v := data.([]uint16)
			ret := make([][][]uint16, dimLengths[0])
			for i := 0; i < int(dimLengths[0]); i++ {
				ret[i] = make([][]uint16,dimLengths[1])
				for j :=0; j< int(dimLengths[1]); j++ {
					ret[i][j]=v[:dimLengths[2]]
					v = v[dimLengths[2]:]
				}
			}
			return ret

		default:
			fail("unknown type", ErrUnknownType)
		}
```

The output of my test program now looks like this:

```
2022-08-03 13:53:15.698  val[0]=[[0 1 2 3] [10 11 12 13] [20 21 22 23]] val[((len(val))-(1))]=[[100 101 102 103] [110 111 112 113] [120 121 122 123]] len(val)=2
```

**Comment by MartinKielhornPhilips** [comment]
Conclusion: The current code base doesn't contain a code path to handle 3D arrays (or higher dimensions). I created a proof-of-concept for the solution for 3D by adapting the 2D code for 3D. I construct a 3D slice in variable `ret` and fill it with consecutive values from the 1D array `v` using two for loops. For a proper patch this code shall be repeated for the other variable types. Extending to more dimensions is straight forward. I suggest to add code up to support 3 to 7 dimensions and would be willing to generate it.

**Comment by MartinKielhornPhilips** [comment]
Maybe there is a codepath with `cdf.slowConvert==true` that would make my example work but I don't know how to set this variable. It is not exposed by the api.Group interface that is returned by cdf.Open.

**Comment by MartinKielhornPhilips** [comment]
The following code handles all types except typeChar for 3D:
```go
	if (!(cdf.slowConvert)) && ((3) == (len(dimLengths))) {
		switch vType {
		case typeByte:
			v := data.([]int8)
			ret := make([][][]int8, dimLengths[0])
			for i := 0; i < int(dimLengths[0]); (i)++ {
				ret[i] = make([][]int8, dimLengths[1])
				for j := 0; j < int(dimLengths[1]); (j)++ {
					ret[i][j] = v[:dimLengths[2]]
					v = v[dimLengths[2]:]
				}
			}
			return ret
		case typeShort:
			v := data.([]int16)
			ret := make([][][]int16, dimLengths[0])
			for i := 0; i < int(dimLengths[0]); (i)++ {
				ret[i] = make([][]int16, dimLengths[1])
				for j := 0; j < int(dimLengths[1]); (j)++ {
					ret[i][j] = v[:dimLengths[2]]
					v = v[dimLengths[2]:]
				}
			}
			return ret
		case typeInt:
			v := data.([]int32)
			ret := make([][][]int32, dimLengths[0])
			for i := 0; i < int(dimLengths[0]); (i)++ {
				ret[i] = make([][]int32, dimLengths[1])
				for j := 0; j < int(dimLengths[1]); (j)++ {
					ret[i][j] = v[:dimLengths[2]]
					v = v[dimLengths[2]:]
				}
			}
			return ret
		case typeFloat:
			v := data.([]float32)
			ret := make([][][]float32, dimLengths[0])
			for i := 0; i < int(dimLengths[0]); (i)++ {
				ret[i] = make([][]float32, dimLengths[1])
				for j := 0; j < int(dimLengths[1]); (j)++ {
					ret[i][j] = v[:dimLengths[2]]
					v = v[dimLengths[2]:]
				}
			}
			return ret
		case typeDouble:
			v := data.([]float64)
			ret := make([][][]float64, dimLengths[0])
			for i := 0; i < int(dimLengths[0]); (i)++ {
				ret[i] = make([][]float64, dimLengths[1])
				for j := 0; j < int(dimLengths[1]); (j)++ {
					ret[i][j] = v[:dimLengths[2]]
					v = v[dimLengths[2]:]
				}
			}
			return ret
		case typeUByte:
			v := data.([]uint8)
			ret := make([][][]uint8, dimLengths[0])
			for i := 0; i < int(dimLengths[0]); (i)++ {
				ret[i] = make([][]uint8, dimLengths[1])
				for j := 0; j < int(dimLengths[1]); (j)++ {
					ret[i][j] = v[:dimLengths[2]]
					v = v[dimLengths[2]:]
				}
			}
			return ret
		case typeUShort:
			v := data.([]uint16)
			ret := make([][][]uint16, dimLengths[0])
			for i := 0; i < int(dimLengths[0]); (i)++ {
				ret[i] = make([][]uint16, dimLengths[1])
				for j := 0; j < int(dimLengths[1]); (j)++ {
					ret[i][j] = v[:dimLengths[2]]
					v = v[dimLengths[2]:]
				}
			}
			return ret
		case typeUInt:
			v := data.([]uint32)
			ret := make([][][]uint32, dimLengths[0])
			for i := 0; i < int(dimLengths[0]); (i)++ {
				ret[i] = make([][]uint32, dimLengths[1])
				for j := 0; j < int(dimLengths[1]); (j)++ {
					ret[i][j] = v[:dimLengths[2]]
					v = v[dimLengths[2]:]
				}
			}
			return ret
		case typeUInt64:
			v := data.([]uint64)
			ret := make([][][]uint64, dimLengths[0])
			for i := 0; i < int(dimLengths[0]); (i)++ {
				ret[i] = make([][]uint64, dimLengths[1])
				for j := 0; j < int(dimLengths[1]); (j)++ {
					ret[i][j] = v[:dimLengths[2]]
					v = v[dimLengths[2]:]
				}
			}
			return ret
		case typeInt64:
			v := data.([]int64)
			ret := make([][][]int64, dimLengths[0])
			for i := 0; i < int(dimLengths[0]); (i)++ {
				ret[i] = make([][]int64, dimLengths[1])
				for j := 0; j < int(dimLengths[1]); (j)++ {
					ret[i][j] = v[:dimLengths[2]]
					v = v[dimLengths[2]:]
				}
			}
			return ret
		default:
			fail("unknown type", ErrUnknownType)
		}
	}
```

**Comment by MartinKielhornPhilips** [comment]
I created the patch for 3 to 7 dimensions and attach them to this comment.

Here is the output of Test.go
```
ubuntu martin@NL:~/stage/kielhorn_dps/67_golang_netcdf_bug/source00$ ./main 
2022-08-03 15:59:37.413 netcdf runtime.Version()=go1.18.3
2022-08-03 15:59:37.413 define coordinate x 
2022-08-03 15:59:37.413 define coordinate y 
2022-08-03 15:59:37.413 define coordinate z 
2022-08-03 15:59:37.413 define coordinate w 
2022-08-03 15:59:37.413 fill 4d dataset val 
2022-08-03 15:59:37.413 close fn=newdata.nc
2022-08-03 15:59:37.413 obtain and show variable 'x' from netcdf file 
2022-08-03 15:59:37.413  x[0]=0 x[((len(x))-(1))]=1 len(x)=2
2022-08-03 15:59:37.413 obtain and show variable 'y' from netcdf file 
2022-08-03 15:59:37.413  y[0]=0 y[((len(y))-(1))]=1 len(y)=2
2022-08-03 15:59:37.413 obtain and show variable 'z' from netcdf file 
2022-08-03 15:59:37.413  z[0]=0 z[((len(z))-(1))]=1 len(z)=2
2022-08-03 15:59:37.413 obtain and show variable 'w' from netcdf file 
2022-08-03 15:59:37.413  w[0]=0 w[((len(w))-(1))]=1 len(w)=2
2022-08-03 15:59:37.413 obtain and show variable 'val' from netcdf file 
2022-08-03 15:59:37.413  val[0]=[[[0 1] [10 11]] [[100 101] [110 111]]] val[((len(val))-(1))]=[[[1000 1001] [1010 1011]] [[1100 1101] [1110 1111]]] len(val)=2
2022-08-03 15:59:37.414 close file fn=newdata.nc
ubuntu martin@NL:~/stage/kielhorn_dps/67_golang_netcdf_bug/source00$ ncdump newdata.nc 
netcdf newdata {
dimensions:
        x = 2 ;
        y = 2 ;
        z = 2 ;
        w = 2 ;
variables:
        ubyte x(x) ;
        ubyte y(y) ;
        ubyte z(z) ;
        ubyte w(w) ;
        uint64 val(x, y, z, w) ;

// global attributes:
data:

 x = 0, 1 ;

 y = 0, 1 ;

 z = 0, 1 ;

 w = 0, 1 ;

 val =
  0, 1,
  10, 11,
  100, 101,
  110, 111,
  1000, 1001,
  1010, 1011,
  1100, 1101,

  1110, 1111 ;
}

```

It seems to work with low numbers but writing seems to fail once the data gets bigger.

Note: I had to rename the go files to .txt extension for github to take them as attachments.

[Test_go.txt](https://github.com/batchatco/go-native-netcdf/files/9252020/Test_go.txt)

[cdfPatch_go.txt](https://github.com/batchatco/go-native-netcdf/files/9252027/cdfPatch_go.txt)

**Comment by MartinKielhornPhilips** [comment]
I used Common Lisp with my library https://github.com/plops/cl-golang-generator to create the code. For completeness I attach the generator code (gen01.lisp) here.

I can understand that no maintainer would want to add 30kB of autogenerated code into their project but I would really appreciate if at least the 3D case can be handled.

[gen01_lisp.txt](https://github.com/batchatco/go-native-netcdf/files/9252157/gen01_lisp.txt)

**Comment by batchatco** [comment]
Should be fixed with 807bf353fe67dec43e817615c8ffc4c3506a6083

Please verify when you get a chance.

**Comment by MartinKielhornPhilips** [comment]
Test with 3D varying dimensions in 4D dataset
(x,y,z,w) = (2,3,4,1)

also looks good:

```
67_golang_netcdf_bug/source00$ ./main
2022-08-22 11:10:13.126 name of program: netcdf; golang compiler version: runtime.Version()=go1.18.3
2022-08-22 11:10:13.126 dependency dep.Path=github.com/batchatco/go-native-netcdf dep.Version=v0.0.0-20220806211823-807bf353fe67 dep.Sum=h1:ETvGOPEyAx7UqU5gEo9zCAjr4tCuOcKJjYAc11J009E=
2022-08-22 11:10:13.126 dependency dep.Path=github.com/batchatco/go-thrower dep.Version=v0.0.0-20200827035905-5cb7337f6be6 dep.Sum=h1:gDf4IUqKDnH7F0XdgeYOBx2jlMKF/j9Xm42sISXpwqY=
2022-08-22 11:10:13.126 dependency dep.Path=github.com/samber/lo dep.Version=v1.27.0 dep.Sum=h1:GOyDWxsblvqYobqsmUuMddPa2/mMzkKyojlXol4+LaQ=
2022-08-22 11:10:13.126 dependency dep.Path=golang.org/x/exp dep.Version=v0.0.0-20220303212507-bbda1eaf7a17 dep.Sum=h1:3MTrJm4PyNL9NBqvYDSj3DHl46qQakyfqfWo4jgfaEM=
2022-08-22 11:10:13.126 define coordinate x 
2022-08-22 11:10:13.126 define coordinate y 
2022-08-22 11:10:13.126 define coordinate z 
2022-08-22 11:10:13.126 define coordinate w 
2022-08-22 11:10:13.126 fill 4d dataset val 
2022-08-22 11:10:13.126 close fn=newdata.nc
2022-08-22 11:10:13.126 obtain and show variable 'x' from netcdf file 
2022-08-22 11:10:13.126  x[0]=0 x[((len(x))-(1))]=1 len(x)=2
2022-08-22 11:10:13.126 obtain and show variable 'y' from netcdf file 
2022-08-22 11:10:13.126  y[0]=0 y[((len(y))-(1))]=2 len(y)=3
2022-08-22 11:10:13.126 obtain and show variable 'z' from netcdf file 
2022-08-22 11:10:13.126  z[0]=0 z[((len(z))-(1))]=3 len(z)=4
2022-08-22 11:10:13.126 obtain and show variable 'w' from netcdf file 
2022-08-22 11:10:13.126  w[0]=0 w[((len(w))-(1))]=0 len(w)=1
2022-08-22 11:10:13.126 obtain and show variable 'val' from netcdf file 
2022-08-22 11:10:13.126  val[0]=[[[0] [10] [20] [30]] [[100] [110] [120] [130]] [[200] [210] [220] [230]]] val[((len(val))-(1))]=[[[1000] [1010] [1020] [1030]] [[1100] [1110] [1120] [1130]] [[1200] [1210] [1220] [1230]]] len(val)=2
2022-08-22 11:10:13.126 close file fn=newdata.nc

67_golang_netcdf_bug/source00$ ncdump newdata.nc 
netcdf newdata {
dimensions:
        x = 2 ;
        y = 3 ;
        z = 4 ;
        w = 1 ;
variables:
        ushort x(x) ;
        ushort y(y) ;
        ushort z(z) ;
        ushort w(w) ;
        uint64 val(x, y, z, w) ;

// global attributes:
data:

 x = 0, 1 ;

 y = 0, 1, 2 ;

 z = 0, 1, 2, 3 ;

 w = 0 ;

 val =
  0,
  10,
  20,
  30,
  100,
  110,
  120,
  130,
  200,
  210,
  220,
  230,
  1000,
  1010,
  1020,
  1030,
  1100,
  1110,
  1120,
  1130,
  1200,
  1210,
  1220,
  1230 ;
}

```

**Comment by MartinKielhornPhilips** [comment]
I think your patch works.
Thanks.

**Comment by batchatco** [comment]
Thanks very much for the report and the verification!
